server/src/main.rs

```diff
@@ -0,0 +1,332 @@
+        )
+        .unwrap();
+        let position = params.text_document_position_params.position;
+        let compilation_state = &self.shared_state.lock().await.compilation_state;
+        Ok(
+            get_hover_info(&compilation_state, uri, position).map(|info| Hover {
+                contents: HoverContents::Scalar(MarkedString::String(info)),
+                range: None,
+            }),
+        )
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file opened!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+
+    async fn did_save(&self, params: DidSaveTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+}
+
+impl Backend {
+    async fn handle_file_change(&self, uri: Url) {
+        let (updated_state, options) = self.compile_slice_files().await;
+
+        let mut shared_state_lock = self.shared_state.lock().await;
+
+        shared_state_lock.compilation_state = updated_state;
+        shared_state_lock.compilation_options = options;
+
+        let diagnostics = self
+            .get_diagnostics_for_uri(&uri, &mut shared_state_lock)
+            .await;
+
+        self.client
+            .publish_diagnostics(uri, diagnostics, None)
+            .await;
+    }
+
+    // This will consume all of the diagnostics in the compilation state and return them as LSP diagnostics.
+    // If you need the diagnostics again you will need to recompile.
+    async fn get_diagnostics_for_uri(
+        &self,
+        uri: &Url,
+        shared_state: &mut SharedState,
+    ) -> Vec<Diagnostic> {
+        let compilation_state = &mut shared_state.compilation_state;
+
+        let options = &shared_state.compilation_options;
+        let diagnostics = std::mem::take(&mut compilation_state.diagnostics).into_updated(
+            &compilation_state.ast,
+            &compilation_state.files,
+            options,
+        );
+
+        diagnostics
+            .iter()
+            .filter_map(|d| {
+                d.span()
+                    .and_then(|s| {
+                        // Convert the span file path to a PathBuf
+                        let span_path = PathBuf::from(&s.file);
+
+                        // Convert the URI to a PathBuf, handling URL decoding
+                        let uri_path = uri.to_file_path().ok().map(|p| p.to_path_buf())?;
+
+                        // Check if the paths match
+                        (span_path == uri_path).then(|| d)
+                    })
+                    // Convert to LSP diagnostic
+                    .and_then(|d| try_into_lsp_diagnostic(d))
+            })
+            .collect::<Vec<_>>()
+    }
+
+    async fn compile_slice_files(&self) -> (CompilationState, SliceOptions) {
```

Comment:
Why return SliceOptions here, and make the caller update the options in the shared state? I'm also unsure about using a SharedState to hold both compilation state and compilation options, as they change independently.

We have a separate `root_uri` which seems only used for setting the Slice options references path, maybe we can get rid of it and update the Slice options directly?

Reply:
We need to know what options produced a given compilation state so we always update them at the same time but we could move the updating into the `compile_slice_files` function as we will always want to do that.

Reply:
Can we just pass the options as a parameter?

Reply:
We could but in the future we are going to want to store and work with multiple compilation states so that the language server can work for multiple independent source directories. I this case is is useful to have a linked copy of the options used for creating the compilation state.

Reply:
Passing the options as a parameter shouldn't be a problem. How would slice compile know which source directory to compile if you don't pass the options as a parameter?

Reply:
> Why return SliceOptions here, and make the caller update the options in the shared state?

I agree, seems better to just have this function _do_ the updating itself.
Instead of returning a tuple, and making the caller update it.
Note that _every_ time we call this function we update the shared state.

> I'm also unsure about using a SharedState to hold both compilation state and compilation options, as they change independently.

Personally, I would move everything into a shared state.
So it's easier to reason about when things happen.
Plus, the independence of these values is limited. Changing any of them  (`workspace_uri`, `sources_uri`, or `compilation_options`) should trigger a fresh recompilation (changing `compilation_state`).

Reply:
This would hamper our ability to concurrently handle requests, but I think it will still be more than fast enough for our simple language server.